### DISCUSSION TITLE
Extract Message From JSON

### DISCUSSION THREAD

**Discussion by JJWest12** [nextgenhealthcare/connect#6373] [Q&A]
I have a channel where the inbound message is a JSON object, and in that JSON object there is a node named "message" that contains an HL7v2 message. I am trying to extract the MSH.7 field from the HL7v2 message that is embedded in the JSON but cannot figure out how do to this. Can someone assist?

**Comment by kirbykn2** [comment]
What have you tried to do?  Can you post an example TEST message (no PHI)

On Tue, Dec 10, 2024 at 8:44 AM JJWest12 ***@***.***> wrote:

> I have a channel where the inbound message is a JSON object, and in that
> JSON object there is a node named "message" that contains an HL7v2 message.
> I am trying to extract the MSH.7 field from the HL7v2 message that is
> embedded in the JSON but cannot figure out how do to this. Can someone
> assist?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nextgenhealthcare/connect/discussions/6373>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APRXWDY3JDVBNO2CTNG2O4L2E3V4XAVCNFSM6AAAAABTLGN4VWVHI2DSMVQWIX3LMV43ERDJONRXK43TNFXW4OZXGY2TAMRVGA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Best,

Kirby Knight
| 231.735.4650 | ***@***.***

**Reply by JJWest12** [reply]
Here is a dumbed down version of the message

{
  "key" : "1334455",
  "message_date" : "2022-01-01 03:11:13",
  "message" : "HL7 HERE"
}


What I have tried is capturing the message then trying to convert it to XML has that I can access MSH.7, but my channel maps keep coming back blank...


    msg = new XML (msg['native_message'].toString().replace(/\\r/g, '\n'))
    default xml namespace = "urn:hl7-org:v3"

    var messageDateTime = msg['MSH']['MSH.7']['MSH.7.1'].toString()
	
    var year = messageDateTime.substring(0,5)
    var month = messageDateTime.substring(5,7)

    channelMap.put('year', year)
    channelMap.put('month', month)

**Reply by pacmano1** [reply]
You have to serialize the message to XML first assuming it is ER7 (ER7 is the standard pipe delimited HL7) to reference it as XML.

**Reply by JJWest12** [reply]
@pacmano1 can you assist me with how to do that please?

**Reply by pacmano1** [reply]
adapt https://forums.mirthproject.io/forum/mirth-connect/support/18895-string-to-hl7?p=104370#post104370

And a reminder to check the forums for readers here, there is a lot of great stuff over there.

**Reply by JJWest12** [reply]
@pacmano1 Weird though I am still getting an error saying that it cannot find "MSH" in undefined. I can see that the encoded message looks like XML now, and the XML has an "MSH" node so I am not sure what I am missing here

    msg = SerializerFactory.getSerializer('HL7V2').toXML(msg['native_message'].toString().replace(/\\r/g, '\n'));
    default xml namespace = "urn:hl7-org:v3"

    var messageDateTime = msg['MSH']['MSH.7']['MSH.7.1'].toString()
	
    var year = messageDateTime.substring(0,5); channelMap.put('year', year)
    var month = messageDateTime.substring(5,7); channelMap.put('month', month)

**Reply by pacmano1** [reply]
Stop replacing stuff.  And stop .toString().    Stop setting the name space.

**Reply by pacmano1** [reply]
@JJWest12 don't blindly copy stuff without understanding what it does.   Not giving you are hard time - but your code reflects that you don't understand what the code is doing.

**Reply by JJWest12** [reply]
@pacmano1 I need the replace because of the way the HL7 message in formatted in the JSON. It was "\r" where new lines need to be

**Reply by pacmano1** [reply]
\r is the standard delimiter for the end of a segment, not \n.   In fact \n can be used inside a segment, e.g. in an OBX for text report.

@kirbykn2 already asked you for a sample message.

**Reply by pacmano1** [reply]
@JJWest12, I should have seen this eariler.

```JS
msg = SerializerFactory.getSerializer('HL7V2').toXML(msg['native_message'].toString().replace(/\\r/g, '\n'));
```
instead
``` JS
msg = new XML(SerializerFactory.getSerializer('HL7V2').toXML(msg['native_message'].toString())
```
